File: vm1/alternate-vm.py
#!/usr/bin/env python
# coding: utf-8

import sys

class Stack:
    def __init__(self):
        self.stack = []

    def push(self, value):
        self.stack.append(value)
    
    def pop(self):
        return self.stack.pop()


class Emulator:
    def __init__(self):
        self.stack = Stack()

    def mul(self):
        b = self.stack.pop()
        a = self.stack.pop()
        c = int(a) * int(b)
        self.stack.push(c)

    def add(self):
        b = self.stack.pop()
        a = self.stack.pop()
        c = int(a) + int(b)
        self.stack.push(c)

    def sub(self):
        b = self.stack.pop()
        a = self.stack.pop()
        c = int(a) - int(b)
        self.stack.push(c)

    def set(self, number):
        self.stack.push(number)

    def printn(self):
        a = self.stack.pop()
        print(a)


    def parse(self, command):
        match command.split():

            case ['MUL']:
                self.mul()

            case ['ADD']:
                self.add()

            case ['SUB']:
                self.sub()

            case ['SET', number]:
                self.set(number)

            case ['PRINT']:
                self.printn()

            case ['HALT']:
                sys.exit(0)

            case _:  # catch all
                print(f"{command!r}?")


class Runner:
    def __init__(self):
        self.emu = Emulator()

    def read(self, file):
        with open(file) as f:
            contents = f.readlines()
            print(contents)
            for line in contents:
                command = line.strip()
                self.emu.parse(command)
        f.close()

run = Runner()
run.read('sample.txt')

File: vm1/alternate-vm.php
<?php

class Stack {

    public $top;
    public $stack = array();

    function __construct() {
        $this->top = -1;
    }

    public function push($x) {
        $this->stack[++$this->top] = $x;
    }

    public function pop() {
        if ($this->top < 0){
            echo "error: stack empty";
        } else {
          return $this->stack[$this->top--];
        }
    }
}


class OP {
    const ADD = 1;
    const MUL = 2;
    const SUB = 3;
    const SET = 4;
    const PRNT = 5;
    const HALT = 6;
}

$code = array(
    OP::SET, 33,
    OP::SET, 44,
    OP::ADD,
    OP::PRNT,
    OP::HALT
);


function run($code): void {

    $stk = new Stack();
    $pc = 0;

    do {
        $opcode = $code[$pc];
        $pc = $pc + 1;

        switch ($opcode) {

            case OP::HALT:
                return;

            case OP::SET:
                $v = $code[$pc];
                $stk->push($v);
                break;

            case OP::ADD:
                $b = $stk->pop();
                $a = $stk->pop();
                $stk->push($a + $b);
                break;

            case OP::SUB:
                $b = $stk->pop();
                $a = $stk->pop();
                $stk->push($a - $b);
                break;

            case OP::MUL:
                $b = $stk->pop();
                $a = $stk->pop();
                $stk->push($a * $b);
                break;

            case OP::PRNT:
                $v = $stk->pop();
                echo "{$v}".PHP_EOL;
                break;

            default:
                break;
        } // switch

    } while (TRUE);
}

run($code);
